pallets/xyk-liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,927 @@
+
+			Self::deposit_event(Event::YieldFarmDestroyed {
+				global_farm_id,
+				yield_farm_id,
+				who,
+				asset_pair,
+			});
+
+			Ok(().into())
+		}
+
+		/// Deposit LP shares to a liq. mining.
+		///
+		/// This function transfers LP shares from `origin` to pallet's account and mint nft for
+		/// `origin` account. Minted nft represents deposit in the liq. mining.
+		///
+		/// Parameters:
+		/// - `origin`: account depositing LP shares. This account has to have at least
+		/// `shares_amount` of LP shares.
+		/// - `global_farm_id`: id of global farm to which user wants to deposit LP shares.
+		/// - `yield_farm_id`: id of yield farm to deposit to.
+		/// - `asset_pair`: asset pair identifying LP shares user wants to deposit.
+		/// - `shares_amount`: amount of LP shares user wants to deposit.
+		///
+		/// Emits `SharesDeposited` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::deposit_shares())]
+		pub fn deposit_shares(
+			origin: OriginFor<T>,
+			global_farm_id: GlobalFarmId,
+			yield_farm_id: YieldFarmId,
+			asset_pair: AssetPair,
+			shares_amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let amm_pool_id = Self::ensure_amm(asset_pair)?;
+
+			let amm_share_token = T::AMM::get_share_token(asset_pair);
+
+			ensure!(
+				T::MultiCurrency::ensure_can_withdraw(amm_share_token, &who, shares_amount).is_ok(),
+				Error::<T>::InsufficientAmmSharesBalance
+			);
+
+			let deposit_id = T::LiquidityMiningHandler::deposit_lp_shares(
+				global_farm_id,
+				yield_farm_id,
+				amm_pool_id,
+				shares_amount,
+				Self::get_asset_balance_in_amm_pool,
+			)?;
+
+			Self::lock_lp_tokens(amm_share_token, &who, shares_amount)?;
+			T::NFTHandler::mint_into(&T::NftCollectionId::get(), &deposit_id, &who)?;
+
+			Self::deposit_event(Event::SharesDeposited {
+				global_farm_id,
+				yield_farm_id,
+				who,
+				amount: shares_amount,
+				lp_token: amm_share_token,
+				deposit_id,
+			});
+
+			Ok(())
+		}
+
+		/// This function create yield farm entry for existing deposit. LP shares are not transferred
+		/// and amount of LP shares is based on existing deposit.
+		///
+		/// This function DOESN'T create new deposit.
+		///
+		/// Parameters:
+		/// - `origin`: account depositing LP shares. This account have to have at least
+		/// - `global_farm_id`: global farm identifier.
+		/// - `yield_farm_id`: yield farm identifier redepositing to.
+		/// - `asset_pair`: asset pair identifying LP shares user want to deposit.
+		/// - `deposit_id`: identifier of the AMM pool.
+		///
+		/// Emits `SharesRedeposited` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::redeposit_lp_shares())]
+		pub fn redeposit_lp_shares(
+			origin: OriginFor<T>,
+			global_farm_id: GlobalFarmId,
+			yield_farm_id: YieldFarmId,
+			asset_pair: AssetPair,
+			deposit_id: DepositId,
+		) -> DispatchResult {
+			let owner = Self::ensure_nft_owner(origin, deposit_id)?;
```

Comment:
I would prefer to check signed origin outside the ensure_nft_owner. 

so first check signed origin and pass the account to check the owner instead of origin.